### PR TITLE
feat(stdlib): Add keccak

### LINKED ISSUE #1189: Create Foreign Keccak256 Function
### Problem

We [already have the Keccak256 opcode in ACVM](https://github.com/noir-lang/acvm/pull/91), but we're missing a Noir function to actually call it in Noir programs.

### Proposed solution

Create the function in `stdlib`.

### Additional context

From @vezenovm on Slack re how the function signature could look like:

> I think there were some efficiency benefits to using a `[u64]` however a byte slice is more intuitive and perhaps we can add helper methods to convert between different array types

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by guipublic**
# Related issue(s)

Resolves #1189

# Description

## Summary of changes

Adds Keccak in stdlib and handle Keccak intrinsic function.

## Dependency additions / changes

Needs a bunch of PR to be merge and versions to be updated before this can be tested:
https://github.com/noir-lang/acvm/pull/247
https://github.com/noir-lang/aztec_backend/pull/150
https://github.com/AztecProtocol/barretenberg/pull/393

## Test additions / changes

I will add an integration test once the above PR are published.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

## Documentation needs
- [X] This PR requires documentation updates when merged.

New Keccak function in the stdlib.

### CODE REVIEW

**Comment by TomAFrench** [comment]
This can be tested by updating the `Cargo.toml` file on https://github.com/noir-lang/aztec_backend/pull/150 to point to the head commit of https://github.com/noir-lang/acvm/pull/247 and then updating this PR to use the updated backend commit.

As an example of this see https://github.com/noir-lang/noir/pull/1226/files/d79b7f3a737940a9528753a05e1ed4d530397064

**Comment by kevaundray** [comment]
Waiting for https://github.com/noir-lang/aztec_backend/pull/164 to merge so we can avoid the memory leak

File: flake.lock [hunk 1]
```diff
@@ -10,11 +10,11 @@
...
       },
       "locked": {
-        "lastModified": 1682626614,
-        "narHash": "sha256-TC535frlYLUTDZ2iHAtUtKpMJWngL1MFxzEXhOfeCo8=",
+        "lastModified": 1683235108,
+        "narHash": "sha256-2NN/pb9hwM5/qJ5mpJfGso1H3fc2RRaZ4sZhVPvT0Dw=",
         "owner": "AztecProtocol",
         "repo": "barretenberg",
-        "rev": "7b5598890c1fa4ee485a4a0015fcb23b5389392e",
+        "rev": "e66f1ef38c3c87c223456d8a77878c2bd3d346eb",
```
Comment (phated) [inline flake.lock:17]: Let's update this to `ad615ee7dc931d3dbea041e47c96b9d8dccebf98` which is latest master.

Comment (kevaundray) [reply flake.lock]: Done

File: crates/nargo_cli/Cargo.toml [hunk 2]
```diff
@@ -37,7 +37,7 @@ termcolor = "1.1.2"
 color-eyre = "0.6.2"
 
 # Backends
-acvm-backend-barretenberg = { git = "https://github.com/noir-lang/aztec_backend", rev = "c9fb9e806f1400a2ff7594a0669bec56025220bb", default-features=false }
```
Comment (phated) [inline crates/nargo_cli/Cargo.toml:40]: This has been merged and needs to be changed to `677f10e07011849f8aa0d75fe80390bb3081b1e5`

Comment (kevaundray) [reply crates/nargo_cli/Cargo.toml:40]: Done

**Review by kevaundray — APPROVED** [review]

**Pull request merged**